### PR TITLE
Added AsStream to PipeReader and PipeWriter

### PR DESCRIPTION
**Pull request by davidfowl**
- This adds a new virtual method `AsStream` to PipeReader and PipeWriter to get a read only or write only stream from the PipeReader and PipeWriter
- This introduces a new field on the base types
- Added tests

Some notes:
- ~~I didn't override the Begin/End overloads, I'm not too concerned about those right now. I can do a follow up to use these helpers (https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/Threading/Tasks/TaskToApm.cs).~~
- Dispose is a noop, I decided not to make it call Complete on the underlying `PipeReader`/`PipeWriter` but we can discuss if it makes sense to do that.
- There are a bunch of optimizations that can be done to ReadAsync to avoid the state machine but I didn't attempt those.

Contributes to https://github.com/dotnet/corefx/issues/27246

### CODE REVIEW

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs [hunk 1]
```diff
@@ -48,6 +50,15 @@ public abstract partial class PipeReader
         /// </remarks>
         public abstract void AdvanceTo(SequencePosition consumed, SequencePosition examined);
 
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public virtual Stream AsStream()
+        {
+            return _stream ?? (_stream = new PipeReaderStream(this));
```
Comment (omariom) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReader.cs]:
```suggestion
            return _stream ??= new PipeReaderStream(this));
```

It is C#8 already :sunglasses:

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                _pipeReader.AdvanceTo(consumed);
+            }
+
+            // This is a buggy PipeReader implementation that returns 0 byte reads even though the PipeReader
+            // isn't completed or cancelled
+            ThrowHelper.ThrowInvalidOperationException_InvalidZeroByteRead();
+            return 0;
+        }
+
+        /// <inheritdoc />
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: Why `/// <inheritdoc />` on this, but not on any of the other overrides?  Since this is an internal class, seems like you could just delete this comment.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
+﻿using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO.Pipelines
+{
+    internal class PipeWriterStream : Stream
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: sealed

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            WriteAsync(buffer, offset, count).GetAwaiter().GetResult();
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: Same argument validation comment as on the read side.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            ValueTask<FlushResult> valueTask = _pipeWriter.FlushAsync(cancellationToken);
+
+            return GetFlushResultAsTask(valueTask);
+        }
+
+        private static Task GetFlushResultAsTask(ValueTask<FlushResult> valueTask)
+        {
+            if (valueTask.IsCompletedSuccessfully)
+            {
+                FlushResult result = valueTask.GetAwaiter().GetResult();
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: This can just be:
```C#
FlushResult result = valueTask.Result;
```

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                {
+                    ThrowHelper.ThrowOperationCanceledException_FlushCanceled();
+                }
+
+                return Task.CompletedTask;
+            }
+
+            static async Task AwaitTask(ValueTask<FlushResult> valueTask)
+            {
+                FlushResult result = await valueTask;
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: .ConfigureAwait(false)

File: src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs [hunk 7]
```diff
@@ -60,6 +60,18 @@ internal static class ThrowHelper
...
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static Exception CreateOperationCanceledException_ReadCanceled() => new OperationCanceledException(SR.ReadCanceledOnPipeReader);
+
+        public static void ThrowOperationCanceledException_FlushCanceled() => throw CreateOperationCanceledException_FlushCanceled();
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static Exception CreateOperationCanceledException_FlushCanceled() => new OperationCanceledException(SR.FlushCanceledOnPipeWriter);
+
+        public static void ThrowInvalidOperationException_InvalidZeroByteRead() => throw CreateInvalidOperationException_InvalidZeroByteRead();
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static Exception CreateInvalidOperationException_InvalidZeroByteRead() => new InvalidOperationException(SR.InvalidZeroByteRead);
```
Comment (stephentoub) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs]: Are we still doing this two-layer pattern because of netfx?

**Comment by stephentoub** [comment]
> I didn't override the Begin/End overloads, I'm not too concerned about those right now.

I'd suggest just adding them in now based on the TaskToApm helper; they end up being one-liners, and if someone happened to use them (most likely because they were going through some library that used them), you'd keep them on the happy path of async rather than async-over-sync-over-async.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                }
+
+                if (result.IsCompleted)
+                {
+                    return 0;
+                }
+            }
+            finally
+            {
+                _pipeReader.AdvanceTo(consumed);
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: Right it’s about not leaving the reader in that “I’m still reading” state. Some of that has since been relaxed where AdvanceTo can be skipped if the next call is Complete.

But it doesn’t look like anything should really throw here. Even if it did, it’s possible you don’t want to make it consume everything unless it returned from the branch where data is returned.

Tl;DR this code is old and was tweaked a bunch of times and maybe there’s some paranoia in here that may no longer be strictly required

File: src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs (hunk 7 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs]: Not 100% sure @pakrym do you remember? It may have been needed in some case but not others.

**Comment by davidfowl** [comment]
> I'd suggest just adding them in now based on the TaskToApm helper; they end up being one-liners, and if someone happened to use them (most likely because they were going through some library that used them), you'd keep them on the happy path of async rather than async-over-sync-over-async.

Begin/End isn't a happy path 😄 . I'll make the changes

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs (hunk 4 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: Gonna leave it relying on the Memory ctor

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        public override long Seek(long offset, SeekOrigin origin) => throw new NotSupportedException();
+
+        public override void SetLength(long value) => throw new NotSupportedException();
+
+        public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
```
[the remark this answers is not in the document]
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: I think it's fine relying on the memory ctor.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 10]
```diff
@@ -0,0 +1,110 @@
+﻿using System;
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: Nit copyright

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+            ReadOnlySequence<byte> sequence = result.Buffer;
+            long bufferLength = sequence.Length;
+            SequencePosition consumed = sequence.Start;
+
+            try
+            {
+                if (bufferLength != 0)
+                {
+                    int actual = (int)Math.Min(bufferLength, buffer.Length);
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: Why cast to int? Just keep it as a long as you compare to buffer length afterwards.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+                    return 0;
+                }
+            }
+            finally
+            {
+                _pipeReader.AdvanceTo(consumed);
+            }
+
+            // This is a buggy PipeReader implementation that returns 0 byte reads even though the PipeReader
+            // isn't completed or canceled
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: Should we fix this in 3.0 then?

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs [hunk 13]
```diff
@@ -0,0 +1,99 @@
+﻿using System.Diagnostics;
```
Comment (jkotalik) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeWriterStream.cs]: Nit: copyright

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs (hunk 12 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: There's nothing to fix. The method is virtual on an abstract class,  if the implementation is buggy, this will throw. Our implementations AFAIK aren't buggy.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs (hunk 11 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/PipeReaderStream.cs]: The return value is an int so it has to be an int eventually.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs (hunk 7 above)
Comment (pakrym) [reply src/System.IO.Pipelines/src/System/IO/Pipelines/ThrowHelper.cs]: I think that's why we started doing it. Pipelines are netstandard1.3 so we tried to optimize for netfx too.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+            await pipe.Writer.WriteAsync(helloBytes);
+            pipe.Writer.Complete();
+
+            Stream stream = pipe.Reader.AsStream();
+
+            var buffer = new byte[5];
+            int read = await stream.ReadAsync(buffer);
+
+            Assert.Equal(5, read);
+            Assert.Equal(helloBytes.AsSpan(0, 5).ToArray(), buffer.AsSpan(0, read).ToArray());
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Assert.Equal should support IEnumerable arguments.

Comment (stephentoub) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Spans aren't IEnumerable.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs [hunk 15]
```diff
@@ -0,0 +1,306 @@
...
+        [Fact]
+        public async Task EndOfPipeReaderReturnsZeroBytesFromReadAsync()
+        {
+            var pipe = new Pipe();
+            pipe.Writer.Complete();
+
+            Stream stream = pipe.Reader.AsStream();
+
+            var buffer = new byte[5];
+            var read = await stream.ReadAsync(buffer);
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Read twice here. We had a bug with it working only once.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 14 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: It shit the bed for Span.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 15 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Read what twice? From the Stream?

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+
+namespace System.IO.Pipelines.Tests
+{
+    public class PipeReaderStreamTests
+    {
+        public delegate Task<int> ReadAsyncDelegate(Stream stream, byte[] data);
+
+        [Theory]
+        [MemberData(nameof(ReadCalls))]
+        public async Task ReadingFromPipeReaderStreamReadsFromUnderlyingPipeReader(ReadAsyncDelegate readAsync)
```
Comment (pakrym) [inline src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: You don't have tests that go async. Do we want any additional coverage in that area?

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 14 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Neither Span and Memory are IEnumerable

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 15 above)
Comment (pakrym) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Yes, past the end of the stream twice.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 14 above)
Comment (pakrym) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Right, my bad.

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 16 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Added

File: src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs (hunk 15 above)
Comment (davidfowl) [reply src/System.IO.Pipelines/tests/PipeReaderStreamTests.nonnetstandard.cs]: Ah, I'll write a different test. I want to verify that I can interleave reading from the stream and pipe.
